Add tool for migrating from local deep storage/Derby metadata

jon-wei opened the pull request:
This PR adds a new tool under `services` meant to assist with the following use case:

Users sometimes begin evaluating Druid with a simple deployment that uses local deep storage and Derby. After using the evaluation deployment for some time and ingesting some segments, they wish to move to MySQL or PostgreSQL and/or use a different deep storage, while keeping their old segments and ingestion setup.

This tool exports the contents of the following Druid tables:
- segments
- rules
- config
- datasource
- supervisors

These tables are chosen since they contain non-transient (like task locks), non-historical entities (like task logs), as the tool is intended for migrations where the user shuts the entire cluster down.

The tool also allows users to specify a new S3 bucket/key, HDFS path, or new local filesystem path, and the entries from the segments table will be rewritten with new load specs. This is to assist with deep storage migration.

Currently only migration from local deep storage combined with Derby metadata is supported (the use case described above), the tool could be later expanded to handle other use cases.

jihoonson commented on docs/content/operations/quickstart-migration.md:
````diff
@@ -0,0 +1,169 @@
+To use the tool, you can run the following command:
+
+```bash
+java -classpath "lib/*:conf/druid/single-server/micro-quickstart/_common" org.apache.druid.cli.Main tools export-metadata -o /tmp/csv
````
Maybe adding the below would make clear where the current directory is.

```bash
$ cd ${DRUID_ROOT}
```

jon-wei replied:
Added `cd ${DRUID_ROOT}`

jihoonson commented on docs/content/operations/quickstart-migration.md:
```diff
@@ -0,0 +1,169 @@
+
+Example import commands for MySQL and PostgreSQL are shown below.
+
+These example import commands expect `/tmp/csv` and its contents to be accessible from the server. For other options, such as importing from the client filesystem, please refer to the MySQL or PostgreSQL documentation.
```
example -> examples?

jon-wei replied:
The usage here is fine

jihoonson commented on docs/content/operations/quickstart-migration.md:
```diff
@@ -0,0 +1,169 @@
+- rules
+- config
+- datasource
+- supervisors
```
It looks that migrating only these tables would be enough for now, but if we extend this tool to support other types of deep storage in the future, then it would be probably worth to include `audit`, `tasks`, and `tasklogs`.

jon-wei replied:
agree, it would be useful to support those in the future

surekhasaharan commented on docs/content/operations/quickstart-migration.md:
```diff
@@ -0,0 +1,169 @@
+
+This helps users migrate segments stored in local deep storage to HDFS.
+
+`--hadoopStorageDirectory`, `h`: The HDFS path that will hold the migrated segments
```
`-h` instead of `h` ?

jon-wei replied:
Fixed

fjy commented on docs/content/operations/quickstart-migration.md:
```diff
@@ -0,0 +1,169 @@
+---
+layout: doc_page
+title: "Migrating Derby Metadata and Local Deep Storage"
```
you can have a non-HA cluster with non local deep storage and derby metadata store

I think the metadata and deep storage migration should have separate docs

fjy replied:
Some food for thought:
when a user goes from single server to cluster, they should first read the deep storage migration doc

when a user goes from non-HA cluster to HA-cluster, they should read the metadata store migration doc

jon-wei replied:
Hm, there may be some redundancy in contents but I'll look into splitting

fjy replied:
Redundancy is fine as long as it is clear to the user what they should do.

jon-wei replied:
I made a separate page for the `export-metadata` tool and made separate pages for metadata/deep storage migration which reference the tool doc page

jon-wei commented:
@jihoonson 

> Also I tested the command in the doc, it emitted the below error. Would you check it please?

I fixed this, I had simplified the example command too much and left out `-Ddruid.extensions.loadList=[]`, thanks

jihoonson commented on docs/content/operations/export-metadata.md:
````diff
@@ -0,0 +1,200 @@
+
+```bash
+cd ${DRUID_ROOT}
+java -classpath "lib/*" -Dlog4j.configurationFile=conf/druid/cluster/_common/log4j2.xml -Ddruid.extensions.directory="extensions" -Ddruid.extensions.loadList=[] org.apache.druid.cli.Main tools export-metadata --connectURI "jdbc:derby://localhost:1527/var/druid/metadata.db;" -o /tmp/csv
````
Would you please add `mkdir /tmp/csv` too? Looks like the output directory must exist before running this command.

jon-wei replied:
Added that and a note about making sure the directory exists
